demos/coreMQTT_Agent/simple_sub_pub_demo.c

```diff
@@ -0,0 +1,560 @@
+     * initiated the subscribe can check the operation's status.  Also send the
+     * status as the notification value.  These things are just done for
+     * demonstration purposes. */
+    pxCommandContext->xReturnStatus = pxReturnInfo->returnCode;
+
+    /* Check if the subscribe operation is a success. Only one topic is
+     * subscribed by this demo. */
+    if( pxReturnInfo->returnCode == MQTTSuccess )
+    {
+        /* Add subscription so that incoming publishes are routed to the application
+         * callback. */
+        xSubscriptionAdded = addSubscription( ( SubscriptionElement_t * ) xGlobalMqttAgentContext.pIncomingCallbackContext,
+                                              pxSubscribeArgs->pSubscribeInfo->pTopicFilter,
+                                              pxSubscribeArgs->pSubscribeInfo->topicFilterLength,
+                                              prvIncomingPublishCallback,
+                                              NULL );
+
+        if( xSubscriptionAdded == false )
+        {
+            LogError( ( "Failed to register an incoming publish callback for topic %.*s.",
+                        pxSubscribeArgs->pSubscribeInfo->topicFilterLength,
+                        pxSubscribeArgs->pSubscribeInfo->pTopicFilter ) );
+        }
+    }
+
+    xTaskNotify( pxCommandContext->xTaskToNotify,
+                 ( uint32_t ) ( pxReturnInfo->returnCode ),
+                 eSetValueWithOverwrite );
+}
+
+/*-----------------------------------------------------------*/
+
+static BaseType_t prvWaitForCommandAcknowledgment( uint32_t * pulNotifiedValue )
+{
+    BaseType_t xReturn;
+
+    /* Wait for this task to get notified, passing out the value it gets
+     * notified with. */
+    xReturn = xTaskNotifyWait( 0,
+                               0,
+                               pulNotifiedValue,
+                               pdMS_TO_TICKS( mqttexampleMS_TO_WAIT_FOR_NOTIFICATION ) );
+    return xReturn;
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvIncomingPublishCallback( void * pvIncomingPublishCallbackContext,
+                                        MQTTPublishInfo_t * pxPublishInfo )
+{
+    static char cTerminatedString[ mqttexampleSTRING_BUFFER_LENGTH ];
+
+    ( void ) pvIncomingPublishCallbackContext;
+
+    /* Create a message that contains the incoming MQTT payload to the logger,
+     * terminating the string first. */
+    if( pxPublishInfo->payloadLength < mqttexampleSTRING_BUFFER_LENGTH )
+    {
+        memcpy( ( void * ) cTerminatedString, pxPublishInfo->pPayload, pxPublishInfo->payloadLength );
+        cTerminatedString[ pxPublishInfo->payloadLength ] = 0x00;
+    }
+    else
+    {
+        memcpy( ( void * ) cTerminatedString, pxPublishInfo->pPayload, mqttexampleSTRING_BUFFER_LENGTH );
+        cTerminatedString[ mqttexampleSTRING_BUFFER_LENGTH - 1 ] = 0x00;
+    }
+
+    LogInfo( ( "Received incoming publish message %s", cTerminatedString ) );
```

Comment:
This is what it looks like on the console:
```
121 25364 [SubPub1] [INFO] Sending publish request to agent with message "Publisher1 publishing message 9" on topic "/filter/Publisher1"
122 25364 [SubPub1] [INFO] Task Publisher1 waiting for publish 9 to complete.
123 26365 [MQTT Agent] [INFO] Received incoming publish message Publisher1 publishing message 9
124 26365 [SubPub1] [INFO] Publish 9 completed successfully.
```

Including the topic doesn't give you any information that isn't already in the payload (i.e. that it's from Publisher1). And since these buffers aren't NULL terminated, I think adding another buffer just to print the topic uses up stack space unnecessarily.

Reply:
My main reason for suggesting logging of the incoming publish topic is to make it visually matching in the logs that the incoming publish topic matches the outgoing one.